File: 프로그래머스/3/81303. 표 편집/표 편집.py
class Node:
    def __init__(self):
        self.removed = False
        self.prev = None
        self.next = None


def solution(n,k,cmds):
    nodes = [Node() for _ in range(n)]
    curr = nodes[k]
    stack = []
    answer = []
    for idx in range(1, n):
        nodes[idx-1].next = nodes[idx]
        nodes[idx].prev = nodes[idx-1]

    for cmd in cmds:
        if cmd[0]=="D":
            step = int(cmd[2:])
            for _ in range(step):
                curr = curr.next
        elif cmd[0]=="U":
            step = int(cmd[2:])
            for _ in range(step):
                curr = curr.prev
        elif cmd[0]=="C":
            stack.append(curr)
            curr.removed = True
            up = curr.prev
            down = curr.next
            if up:
                up.next = down
            if down:
                down.prev = up
                curr = down
            else:
                curr = up
        else:
            node = stack.pop()
            node.removed = False
            up = node.prev
            down = node.next
            if up:
                up.next = node
            if down:
                down.prev = node
            
                
    
    for i in range(len(nodes)):
        if nodes[i].removed:
            answer.append("X")
        else:
            answer.append("O")
    
    return "".join(answer)

File: 프로그래머스/3/81303. 표 편집/표 편집.js
class Node {
  constructor() {
    this.removed = false;
    this.prev = null;
    this.next = null;
  }
}

function solution(n, k, cmds) {
  const nodes = Array.from({ length: n }, () => new Node());
  let curr = nodes[k];
  const stack = [];

  // 연결 리스트 구성
  for (let i = 1; i < n; i++) {
    nodes[i - 1].next = nodes[i];
    nodes[i].prev = nodes[i - 1];
  }

  for (const cmd of cmds) {
    const [c, num] = cmd.split(" ");

    if (c === "D") {
      let step = Number(num);
      for (let i = 0; i < step; i++) curr = curr.next;
    } 
    else if (c === "U") {
      let step = Number(num);
      for (let i = 0; i < step; i++) curr = curr.prev;
    } 
    else if (c === "C") {
      stack.push(curr);
      curr.removed = true;

      const up = curr.prev;
      const down = curr.next;

      if (up) up.next = down;
      if (down) {
        down.prev = up;
        curr = down;
      } else {
        curr = up;
      }
    } 
    else if (c === "Z") {
      const item = stack.pop();
      item.removed = false;

      const up = item.prev;
      const down = item.next;

      if (up) up.next = item;
      if (down) down.prev = item;
    }
  }

  return nodes.map(node => (node.removed ? "X" : "O")).join("");
}
